pkg/controller/helper/shared.go

```diff
@@ -0,0 +1,93 @@
+package helper
+
+import (
+	"context"
+
+	capabilitiesv1alpha1 "github.com/3scale/3scale-operator/apis/capabilities/v1alpha1"
+	
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+)
+
+/*
+RetrieveTenantCR retrieves tenantCR of a tenant that matches the provider account org name
+- providerAccount
+- k8client
+- logger
+If the tenantList is empty it will return nil, nil
+If tenantCR for given providerAccount org is not present, it will return nil, nil
+*/
+func RetrieveTenantCR(providerAccount *ProviderAccount, client k8sclient.Client, logger logr.Logger, namespace string) (*capabilitiesv1alpha1.Tenant, error) {
+	// Retrieve all product CRs that are under the same ns as the backend CR
+	opts := k8sclient.ListOptions{
+		Namespace: namespace,
+	}
+	
+	tenantList := &capabilitiesv1alpha1.TenantList{}
+	err := client.List(context.TODO(), tenantList, &opts)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, tenant := range tenantList.Items {
+		tenantSecret, err := retrieveTenantSecret(client, &tenant)
+		if err != nil {
+			return nil, err
+		}
+
+		adminURL, ok := tenantSecret.Data[string("adminURL")]
+		if !ok {
+			return nil, err
+		}
+
+		if string(adminURL) == providerAccount.AdminURLStr {
+			return &tenant, nil
+		}
+	}
+
+	return nil, nil
+}
+
+/*
+SetOwnersReference sets ownersReference in given object
+- object
+- k8client
+- tenantCR
+*/
+func SetOwnersReference(object controllerutil.Object, client k8sclient.Client, tenantCR *capabilitiesv1alpha1.Tenant) error {
+	ownerReference := []metav1.OwnerReference{
+		{
+			APIVersion: tenantCR.APIVersion,
+			Kind:       tenantCR.Kind,
+			Name:       tenantCR.Name,
+			UID:        tenantCR.UID,
+		},
+	}
+
+	object.SetOwnerReferences(ownerReference)
```

Comment:
When the object has already the owner reference to the parent tenant, how does this method behave?

In the controller reconciliation pattern, you always need to check if the action to be taken has already been done.

Reply:
Thanks, from the notes on this function:
```
SetOwnerReference is a helper method to make sure the given object contains an object reference to the object provided. This allows you to declare that owner has a dependency on the object without specifying it as a controller. If a reference to the same object already exists, it'll be overwritten with the newly provided version. 
```
So it will be overwritten. 
Would you rather see this check if present and only run when not present?

Reply:
IMO, I would leave it as it is not harming us in any way.

Reply:
Thanks, going to work on an update asap